main/apps.js

```diff
@@ -44,40 +44,91 @@ const config = require('./config');
 const registryApi = require('./registryApi');
 const fileUtil = require('./fileUtil');
 const net = require('./net');
+const settings = require('./settings');
+
+/**
+ * Create sources.json if it does not exist.
+ *
+ * @returns {Promise} promise that resolves if successful.
+ */
+function createSourcesJsonIfNotExists() {
+    const sources = settings.getSources();
+    settings.setSources(sources);
+    return Promise.resolve(sources);
+}
 
 /**
  * Create apps.json if it does not exist.
  *
+ * @param {string} source name of external source.
  * @returns {Promise} promise that resolves if successful.
  */
-function createAppsJsonIfNotExists() {
-    return fileUtil.createJsonFileIfNotExists(config.getAppsJsonPath(), {});
+function createAppsJsonIfNotExists(source) {
+    return fileUtil.createJsonFileIfNotExists(config.getAppsJsonPath(source), {});
 }
 
 /**
  * Create updates.json if it does not exist.
  *
+ * @param {string} source name of external source
  * @returns {Promise} promise that resolves if successful.
  */
-function createUpdatesJsonIfNotExists() {
-    return fileUtil.createJsonFileIfNotExists(config.getUpdatesJsonPath(), {});
+function createUpdatesJsonIfNotExists(source) {
+    return fileUtil.createJsonFileIfNotExists(config.getUpdatesJsonPath(source), {});
+}
+
+/**
+ * Initialize directory structure of a source under external apps.
+ *
+ * @param {string} source name of external source.
+ * @returns {Promise} promise that resolves is successful.
+ */
+function initSourceDirectory(source) {
+    return fileUtil.mkdirIfNotExists(config.getAppsRootDir(source))
+        .then(() => fileUtil.mkdirIfNotExists(config.getNodeModulesDir(source)))
+        .then(() => createAppsJsonIfNotExists(source))
+        .then(() => createUpdatesJsonIfNotExists(source));
 }
 
 /**
  * Download the apps.json file containing a list of official apps. The file
  * is downloaded from the configured url and written to the apps root dir.
  *
+ * @param {string} appsJsonUrl URL of apps.json of external source
  * @returns {Promise} promise that resolves if successful.
  */
-function downloadAppsJsonFile() {
-    return net.downloadToFile(config.getAppsJsonUrl(), config.getAppsJsonPath())
+function downloadAppsJsonFile(appsJsonUrl) {
+    return net.downloadToJson(appsJsonUrl)
         .catch(error => {
             throw new Error(`Unable to download apps list: ${error.message}. If you ` +
                 'are using a proxy server, you may need to configure it as described on ' +
                 'https://github.com/NordicSemiconductor/pc-nrfconnect-core');
+        })
+        .then(appsJson => {
+            const source = appsJson._source; // eslint-disable-line
```

Comment:
Is it possible to remove '_'?

Reply:
The underscore comes from Json which is intentional as a meta tag, I'll comment about it.